File: Sorting algos/Insertion_sort/InsertionSort.scala
object InsertionSort {
  def main(args: Array[String]): Unit = {
    val elements = List(9, 8, 7, 6, 5, 4, 3, 2, 1)
    println(isort(elements).toString)
  }

  def isort(xs: List[Int]): List[Int] = {
    if(xs.isEmpty) Nil
    else insert(xs.head, isort(xs.tail))
  }

  def insert(x: Int, xs: List[Int]): List[Int] = {
    if(xs.isEmpty || x <= xs.head) x :: xs
    else xs.head :: insert(x, xs.tail)
  }
}

File: Sorting algos/Insertion_sort/InsertionSort.java
class InsertionSort {
    public static void main(String[] args) {
        int[] array = new int[] {6, 5, 4, 3, 2, 1};
        printArray(array);
        insertionSort(array);
        printArray(array);
    }

    public static void insertionSort(int[] array) {
        for(int i = 1; i < array.length; i++) {
            int key = array[i];
            int j = i - 1;
            while(j >= 0 && key < array[j]) {
                array[j+1] = array[j];
                j--;
            }
            array[j+1] = key;
        }
    }

    private static void printArray(int[] array) {
        System.out.print("Array: ");
        for(int i = 0; i < array.length; i++) {
            System.out.print(array[i] + " ");
        }
        System.out.println();
    }
}